File: 2522-partition-string-into-substrings-with-values-at-most-k/2522-partition-string-into-substrings-with-values-at-most-k.py
class Solution:
    def __init__(self):
        self.maxcount = float('inf')

    def recurse(self, s, start, end, k, count):
        if end >= len(s):
            self.maxcount = min(self.maxcount, count)
            return

        subst = s[start:end + 1]
        digits = str(k)

        if len(subst) <= len(digits) and int(subst) <= k:
            self.recurse(s, start, end + 1, k, count)
        else:
            subst = s[end]
            if int(subst) > k:
                return
            self.recurse(s, end, end + 1, k, count + 1)

    def minimumPartition(self, s, k):
        self.recurse(s, 0, 0, k, 1)
        return self.maxcount if self.maxcount != float('inf') else -1

File: 2522-partition-string-into-substrings-with-values-at-most-k/2522-partition-string-into-substrings-with-values-at-most-k.java
class Solution {
    int maxcount = Integer.MAX_VALUE;
    
    public void recurse(String str, int start, int end, int k, int count) {
        if(end>= str.length())
        {
            maxcount = Math.min(maxcount, count);
            return;
        }
        String subst = str.substring(start, end+1);
        String digits=Integer.toString(k);
        if(subst.length() <= digits.length() && Integer.parseInt(subst) <= k) {
            recurse(str, start, end + 1, k, count);
        }
        else 
        {
            subst = str.substring(end, end+1);
            if(Integer.parseInt(subst) > k)
                return;
            recurse(str, end, end + 1, k, count + 1);            
        }        
    }
    
    public int minimumPartition(String s, int k) {
        
        recurse(s, 0, 0, k, 1);
        return maxcount == Integer.MAX_VALUE ? -1 : maxcount;
    }
}